File: StreamerLib/AudioEncoder.cs
﻿namespace StreamerLib;

/*
public unsafe struct Ddagrab : IDisposable
{
    public AVInputFormat* inputFormat;
    public AVFormatContext* formatContext;
    public AVCodecParameters* codecParameters;
    public AVCodec* codec;
    public AVCodecContext* codecContext;
    public AVPacket* packet;
    public AVFrame* hwFrame;
    public AVRational timebaseMin;

    public Ddagrab(string ddagrabParameters = "")
    {
        ffmpeg.avdevice_register_all();
        inputFormat = ffmpeg.av_find_input_format("lavfi");
        formatContext = ffmpeg.avformat_alloc_context();
        codecParameters = null;
        codec = null;
        codecContext = null;
        packet = ffmpeg.av_packet_alloc();
        hwFrame = ffmpeg.av_frame_alloc();
        
        fixed(AVFormatContext** fc = &formatContext)
            ffmpeg.avformat_open_input(fc, $"ddagrab={ddagrabParameters}", inputFormat, null);
        ffmpeg.avformat_find_stream_info(formatContext, null);
        codecParameters = formatContext->streams[0]->codecpar;
        codec = ffmpeg.avcodec_find_decoder(codecParameters->codec_id);
        codecContext = ffmpeg.avcodec_alloc_context3(codec);
        ffmpeg.avcodec_parameters_to_context(codecContext, codecParameters);
        ffmpeg.avcodec_open2(codecContext, codec, null);

        ffmpeg.av_read_frame(formatContext, packet);
        ffmpeg.avcodec_send_packet(codecContext, packet);
        ffmpeg.avcodec_receive_frame(codecContext, hwFrame);
        timebaseMin.num = formatContext->streams[0]->avg_frame_rate.den;
        timebaseMin.den = formatContext->streams[0]->avg_frame_rate.num;
    }
    
    /// <summary>
    /// read, decode and write AVFrame to this.hwFrame
    /// </summary>
    /// <returns>this.hwFrame</returns>
    public AVFrame* ReadAvFrame()
    {
        ffmpeg.av_read_frame(formatContext, packet);
        ffmpeg.avcodec_send_packet(codecContext, packet);
        ffmpeg.avcodec_receive_frame(codecContext, hwFrame);
        
        return hwFrame;
    }

    public void Dispose()
    {
        if (formatContext != null)
            ffmpeg.avformat_free_context(formatContext);
        
        if (codecParameters != null)
            fixed(AVCodecParameters** p = &codecParameters)
                ffmpeg.avcodec_parameters_free(p);
        
        if (packet != null || codecContext != null)
        {
            ffmpeg.av_packet_unref(packet);
            ffmpeg.avcodec_send_packet(codecContext, packet);//flush codecContext
        }
        
        if (codecContext != null)
            fixed (AVCodecContext** p = &codecContext)
                ffmpeg.avcodec_free_context(p);
        
        if (packet != null)
            fixed(AVPacket** p = &packet)
                ffmpeg.av_packet_free(p);
        
        if (hwFrame != null)
            fixed(AVFrame** p = &hwFrame)
                ffmpeg.av_frame_free(p);
    }
}
*/

/*
public unsafe struct HardwareEncoder : IDisposable
{
    public AVCodec* codec;
    public AVCodecParameters* codecParameters;
    public AVCodecContext* codecContext;
    public AVPacket* packet;
    public int StreamIndex;
    private StreamWriter _streamWriter;
    private AVRational _timebase;

    public HardwareEncoder(AVFormatContext* formatContext, AVBufferRef* hwFramesContext, string hardwareEncoderName, StreamWriter streamWriter)
    {
        _streamWriter = streamWriter;
        codec = null;
        codecParameters = ffmpeg.avcodec_parameters_alloc();
        codecContext = null;
        packet = ffmpeg.av_packet_alloc();
        
        codec = ffmpeg.avcodec_find_encoder_by_name(hardwareEncoderName);
        codecContext = ffmpeg.avcodec_alloc_context3(codec);
        _timebase = formatContext->streams[0]->time_base;
        codecContext->time_base = _timebase;
        codecContext->pix_fmt = AVPixelFormat.AV_PIX_FMT_D3D11;
        codecContext->width = formatContext->streams[0]->codecpar->width;
        codecContext->height = formatContext->streams[0]->codecpar->height;
        codecContext->max_b_frames = 0;
        codecContext->framerate = ffmpeg.av_guess_frame_rate(null, formatContext->streams[0], null);
        codecContext->hw_frames_ctx = ffmpeg.av_buffer_ref(hwFramesContext);
        ffmpeg.avcodec_open2(codecContext, codec, null);
        ffmpeg.avcodec_parameters_from_context(codecParameters, codecContext);
    }

    public void EncodeAndWriteFrame(AVFrame* hwFrame)
    {
        ffmpeg.avcodec_send_frame(codecContext, hwFrame);
        if (ffmpeg.avcodec_receive_packet(codecContext, packet) != 0)
            return;
        ffmpeg.av_packet_rescale_ts(packet, codecContext->time_base, _streamWriter.GetStreamParameters(StreamIndex).Timebase);
        _streamWriter.WriteFrame(packet, _timebase);
        Console.WriteLine("frame " + codecContext->frame_number + " writed");
    }

    public void Dispose()
    {
        fixed(AVCodecParameters** p = &codecParameters)
            ffmpeg.avcodec_parameters_free(p);
        ffmpeg.av_packet_unref(packet);
        ffmpeg.avcodec_send_packet(codecContext, packet); //flush codecContext
        fixed (AVCodecContext** p = &codecContext)
            ffmpeg.avcodec_free_context(p);
        fixed(AVPacket** p = &packet)
            ffmpeg.av_packet_free(p);
    }
}
*/

public unsafe class AudioEncoder : IDisposable
{
    public const int AV_SAMPLE_FMT_FLT = 3;

    public int SampleSizeInBytes { get; } = 4;
    public int FrameSizeInSamples { get; }
    public int FrameSizeInBytes { get; }
    public int Channels { get; }
    public int StreamIndex { get; private set; } = -1;
    public int SampleRate => _sampleRate;
    public int SampleFormat => AV_SAMPLE_FMT_FLT;

    private IntPtr _codecContext;
    private IntPtr _avFrame;
    private IntPtr _packet;
    private StreamWriter? _streamWriter;
    private long _pts;
    private IntPtr _timebase;
    private IntPtr _codecParameters;
    private int _sampleRate;

    public AudioEncoder(StreamWriter streamWriter, Encoders encoder, int channels = 2)
    {
        string encoderName = encoder switch
        {
            Encoders.LibOpus => "libopus",
            _ => "libopus"
        };

        _sampleRate = 48000; //encoder specified
        Channels = channels;

        int frameSizeInSamples = -1;
        nint codecContext, packet, timebase, codecParameters, frame;

        FFmpegImport.AudioEncoder_Constructor(
            encoderName,
            48000,
            channels,
            &frameSizeInSamples,
            &codecContext,
            &packet,
            &timebase,
            &codecParameters,
            &frame);

        _codecContext = codecContext;
        _packet = packet;
        _timebase = timebase;
        _codecParameters = codecParameters;
        _avFrame = frame;
        
        FrameSizeInSamples = frameSizeInSamples;
        FrameSizeInBytes = frameSizeInSamples * channels * SampleSizeInBytes;
    }

    public void EncodeAndWriteFrame(byte[] buffer)
    {
        if (StreamIndex == -1)
            return;
        if (buffer.Length != FrameSizeInBytes)
            throw new ArgumentException();

        bool success;
        fixed (byte* buf = buffer)
        {
            success = FFmpegImport.AudioEncoder_EncodeAndWriteFrame(
                buf,
                FrameSizeInBytes,
                Channels,
                StreamIndex,
                _pts,
                _codecContext,
                _packet,
                _avFrame);
            if (success)
                _streamWriter.WriteFrame(_packet, _timebase, StreamIndex);
        }

        _pts += FrameSizeInSamples;
    }

    public void Dispose()
    {
        FFmpegImport.AudioEncoder_Dispose(
            _packet, 
            _avFrame, 
            _codecContext, 
            _codecParameters, 
            _timebase);
    }

    public void Clean()
    {
        _pts = 0;
    }

    public void RegisterAVStream(StreamWriter streamWriter)
    {
        _streamWriter = streamWriter;
        StreamIndex = _streamWriter.AddAvStream(_codecParameters, _timebase);
    }
}

/*
public static class FFmpegHelper
{
    public static unsafe void ErrStrPrint(int errNum)
    {
        int maxErrorStringSize = 64;
        byte[] str1 = new byte[maxErrorStringSize];
        fixed (byte* str2 = str1)
        {
            ffmpeg.av_make_error_string(str2, 64, errNum);
            DebugLogUnmanagedPtr(str2);
        }
    }
    
    public static void InicializeFFmpeg()
    {
        SetFfmpegBinaresPath(@"C:\Users\Cray\Desktop\Programs\ffmpeg");
        DynamicallyLoadedBindings.Initialize();
    }

    public static unsafe void DebugLogUnmanagedPtr(byte* ptr) =>
        Debug.WriteLine(Marshal.PtrToStringAnsi((IntPtr)ptr));

    public static unsafe string UnmanagedPtrToString(byte* ptr) =>
        Marshal.PtrToStringAnsi((IntPtr)ptr) ?? String.Empty;

    static void SetFfmpegBinaresPath()=>
        DynamicallyLoadedBindings.LibrariesPath = Path.Combine(Environment.CurrentDirectory, "ffmpeg", "bin");

    static void SetFfmpegBinaresPath(string path) =>
        DynamicallyLoadedBindings.LibrariesPath = path; 
}
*/


//hwframes
//AVBufferRef* hwDeviceContext = null;
//ffmpeg.av_hwdevice_ctx_create(&hwDeviceContext, AVHWDeviceType.AV_HWDEVICE_TYPE_D3D11VA, null, null, 0);
//AVBufferRef* hwFramesRef = ffmpeg.av_hwframe_ctx_alloc(hwDeviceContext);
//AVHWFramesContext* hwFramesContext = (AVHWFramesContext*)(hwFramesRef->data);
//hwFramesContext->format = AVPixelFormat.AV_PIX_FMT_D3D11;
//hwFramesContext->sw_format = AVPixelFormat.AV_PIX_FMT_BGRA;
//hwFramesContext->width = codecContext->width;
//hwFramesContext->height = codecContext->height;
//ffmpeg.av_hwframe_ctx_init(hwFramesRef);

File: libutil/libutil/src/audio_encoder.rs
#![allow(dead_code)]
use core::panic;
use crate::stream_writer::StreamWriter;
use ffmpeg::*;

pub struct AudioEncoder{
    stream_index: i32,
    stream_writer: StreamWriter,
    packet_time_stamp: i64,
    codec_context: *mut AVCodecContext,
    packet: *mut AVPacket,
    frame: *mut AVFrame,
    codec_parameters: *mut AVCodecParameters,
}

impl AudioEncoder{
    pub fn new(audio_codec: AudioCodecs, stream_writer: StreamWriter, channels: i32, required_sample_rate: i32) -> Self{
        let codec_id = match audio_codec {
            AudioCodecs::libopus => {
                if required_sample_rate != 48000 {panic!()}
                AVCodecID_AV_CODEC_ID_OPUS},
            _ => panic!(),
        };
        
        unsafe{
            let packet = av_packet_alloc();

            let codec = avcodec_find_encoder(codec_id);
            let codec_context = avcodec_alloc_context3(codec);
            (*codec_context).sample_rate = required_sample_rate;
            (*codec_context).sample_fmt = AVSampleFormat_AV_SAMPLE_FMT_FLT;
            (*codec_context).time_base = AVRational{ num: 1, den: required_sample_rate};
            av_channel_layout_default(&mut (*codec_context).ch_layout, channels);
            avcodec_open2(codec_context, codec, std::ptr::null_mut());

            let codec_parameters = avcodec_parameters_alloc();
            avcodec_parameters_from_context(codec_parameters, codec_context);
            
            let frame = av_frame_alloc();
            (*frame).nb_samples = (*codec_context).frame_size;
            (*frame).format = AVSampleFormat_AV_SAMPLE_FMT_FLT;
            av_channel_layout_default(&mut (*frame).ch_layout, channels);

            Self { 
                stream_index: -1, 
                stream_writer, 
                packet_time_stamp: 0, 
                codec_context, 
                packet, 
                frame, 
                codec_parameters}
        }
    }
    unsafe fn encode_buffer(&self, buffer: &[f32]) -> i32{
        av_packet_unref(self.packet);
        (*self.frame).pts = self.packet_time_stamp;
        let mut ret = avcodec_fill_audio_frame(self.frame, self.channels(), AVSampleFormat_AV_SAMPLE_FMT_FLT, buffer.as_ptr().cast(), self.frame_size_in_bytes(), self.channels() * 4);
        if ret < 0 {return ret};
        ret = avcodec_send_frame(self.codec_context, self.frame);
        if ret < 0 {return ret};
        ret = avcodec_receive_packet(self.codec_context, self.packet);
        if ret == 0 {
            (*self.packet).stream_index = self.stream_index};
        return ret;
    }

    pub fn register_avstream(&mut self, stream_writer: StreamWriter){
        self.stream_writer = stream_writer;
        self.stream_index = self.stream_writer.add_av_stream(self.codec_parameters)
    }

    pub fn channels(&self) -> i32{
        if self.codec_context.is_null() {
            return 0;
        }
        unsafe {
            (*self.codec_context).channels
        }
    }

    pub fn frame_size_in_bytes(&self) -> i32{
        if self.codec_context.is_null() {
            return 0;
        }
        unsafe {
            (*self.codec_context).frame_size * 4
        }
    }

    pub fn stream_index(&self) -> i32{
        self.stream_index
    }
}

impl Drop for AudioEncoder{
    fn drop(&mut self) {
        unsafe{
            av_packet_unref(self.packet);
            avcodec_send_packet(self.codec_context, self.packet); //flush codec_context
            avcodec_free_context(&mut self.codec_context);
            av_packet_free(&mut self.packet);
            avcodec_parameters_free(&mut self.codec_parameters)
        }
        
    }
}

pub enum AudioCodecs{
    libopus,
}